### PR TITLE
[ARC] Allow retains/releases during deinit

### PR DESCRIPTION
**Pull request by segiddins**

### CODE REVIEW

File: Sources/Runtime/arc.cpp [hunk 1]
```diff
@@ -157,28 +159,58 @@ struct RefCounted {
...
+    box->retainCount = 0; // clear the deallocating flag
+    
     // Explicitly unlock the mutex before deleting the box
     box->mutex.unlock();
 
 #ifdef DEBUG_ARC
     box->deallocated = true;
 #else
-    free(box);
+    delete box;
```
Comment (harlanhaskins) [inline Sources/Runtime/arc.cpp]: Can't delete. Memory wasn't allocated using the standard `new`

**Review by harlanhaskins — APPROVED** [review]

**Review by harlanhaskins — DISMISSED** [review]
Fix the `delete` thing and we're good

File: Sources/Runtime/arc.cpp (hunk 1 above)
Comment (segiddins) [reply Sources/Runtime/arc.cpp]: `delete` seems to work fine on placement operator new... (both according to the docs & the added example)

Comment (harlanhaskins) [reply Sources/Runtime/arc.cpp]: I don't think that's true. C++ has no way of knowing to free the entire buffer that box was allocated into (unless we're relying on an implementation detail of `delete` using free and asking the heap table for the buffer size)

From the googling I've done, it looks like the correct thing here, as we allocated this buffer larger than the object size, is

```c++
box->~RefCountBox();
free(box);
```

**Pull request merged**